### ISSUE TITLE
Support MultiCompilers

### ISSUE THREAD

**Issue by brianmhunt** [shellscape/webpack-plugin-serve#39]
### Feature Use Case

Serving a project that has both `target: 'web'` and `target: 'webworker'` e.g.

#### webpack.config.js

```
module.exports = [
  { 
    entry: 'web.js',
    target: 'web'
  },
  {
    entry: 'worker.js',
    target: 'webworker'
  }
]
```

### Feature Proposal

Documenting how this could might be possible.

**Comment by shellscape** [comment]
@brianmhunt thanks for opening an issue.

I'm somewhat of a web-worker newbie. Your build will still be a webpack build when using this plugin. Are you asking how to have webpack build a web app and webworker bundle separately? If so, that's a question for the main webpack project at https://github.com/webpack/webpack.

If you're asking how you might serve the app and the webworker together using this plugin, then that's our in wheelhouse. I'd need to know a bit more about your implementation, and perhaps get a bare-bones repo that builds a bundle the way your scenario expects.

**Comment by brianmhunt** [comment]
Thanks @shellscape — it's your wheelhouse by the looks of it 😁 

I'll put together a min working repo and report back shortly.

**Comment by brianmhunt** [comment]
I've put together what I think might illustrate the problem, here:

https://github.com/brianmhunt/bug-reports/tree/39-webpack-plugin-serve

Not being familiar with the plugin, I might not be using things right, but I think the crux shown is that the `worker.js` being served is not compiled by webpack, because there are two configs (as `webworker` is a separate target, since regular worker sometimes uses `window` as the global).

**Comment by shellscape** [comment]
Thanks for putting that together. I think we see the larger problems here. One has to do with MultiCompilers. I'll sit down to tackle that tonight.

**Comment by brianmhunt** [comment]
Yeah, I'm sorry to say didn't think it'd be easy, but I thought to pass it along.

Not knowing anything really, I was thinking of a shared `Serve` instance i.e.

```js
const serve = new Serve(options)
module.exports = [{
   plugins: [serve]
}, {
  plugins: [serve]
}]
```

But I really don't know the implications of that *at all*. lol

**Comment by PlayMa256** [comment]
The plugin would not follow the webpack watch lifecycle, we are looking for better ways, thank you for bringing this up!!

**Comment by PlayMa256** [comment]
That would spawn two instances, which is not something we want. Thank you for bringing this up!

**Comment by shellscape** [comment]
@brianmhunt OK so I triaged this, and found a few things that'll allow us to make this work like you were expecting (and like the other dev servers presently work) with a MultiCompiler setup. 

We're going to have to refactor a bit and it's going to involve some testing (always a good thing). I'll probably have a PR ready later tonight and hopefully we can get something ready to go for you tomorrow.

**Comment by shellscape** [comment]
So a quick update: I've got a good method in place to apply the same instance to multiple compilers and let everything work well together. But with that we ended up identifying an issue with the filenames for the metadata that webpack generated (for which we were working around another issue with webpack). So we're trying to figure out a workaround in our workaround hah. 

Down the rabbit hole we go

**Comment by shellscape** [comment]
@brianmhunt have you used the same setup with webpack-dev-server or webpack-serve?

I have this working, but like webpack-serve, HMR doesn't work for the second compiler (in this case the webworker). I haven't actually been able to configure webpack-dev-server to work correctly with my test files to see if it's any different.

It would be very helpful to know if you've used this with another server that successfully applied HMR to both compilers.

**Comment by shellscape** [comment]
good news to share - we've got HMR working with multiple compilers. not sure how that will work with a `webworker` target, but with two `web` targeted compilers, tests show that each compiler can independently apply HMR updates. (this isn't something that webpack-dev-server or webpack-serve can do at present, which is a cool bonus)

there are a few kinks to iron out before we can publish the changes, so please bear with us!

**Comment by brianmhunt** [comment]
No rush - that’s an amazing turnaround 😀

**Comment by shellscape** [comment]
@brianmhunt We're getting there, but there are still some client issues (the build status overlay) that need adjustment before we can merge the fixes. It's a bit complex, and I'd like to have some tests written around it before we release into the wild.

**Comment by brianmhunt** [comment]
No worries at all @shellscape  — it'll be a great addition when it's all set.

**Comment by shellscape** [comment]
@brianmhunt any chance you can try out the `fix/multicompilers` branch? We think we have this ready to go, but I'd like you to verify. Note the `test/fixtures/multi/webpack.config.js` setup. You'll need some slight sugar in your second compiler and you'll have to add the client entry to the second compiler as well.

**Comment by brianmhunt** [comment]
Thanks @shellscape - I haven’t had a chance to check it out yet but I will and I’ll open a new issue & reference this if there are any problems.  Looks great, thanks for doing this, and I’m excited to check it out.
